File: linked-list/easy_difficulty/mid_difficulty/143-reorder-list/reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
     ListNode* reverseList(ListNode* head) {
        if(head == nullptr){
            return nullptr;
        }

        if(head->next == nullptr){
            return head;
        }

        ListNode* crr = head->next;
        ListNode* next = head->next->next;
        ListNode* prev = head;

        head->next = nullptr;
        
        while(next != nullptr){
            crr->next = prev;
            prev = crr;
            crr = next;
            next = next->next;
        }

        crr->next = prev;
        return crr;
    }

    ListNode* middleNode(ListNode* head) {
        if(head == nullptr){
            return nullptr;
        }

        if(head->next == nullptr){
            return head;
        }

        ListNode* middle = head;
        ListNode* fast = head;

        while(fast != nullptr && fast->next != nullptr){
            middle = middle->next;
            fast = fast->next->next;
        }
        return middle;
    }

    void reorderList(ListNode* head) {
        if(head == nullptr){
            return;
        }

        ListNode* middle = reverseList(middleNode(head));

        ListNode* crr = head;
        ListNode* crrNext = head->next;
        ListNode* middleNext = middle->next;

        while(middleNext != nullptr && crrNext != nullptr){
            crr->next = middle;
            crr = crrNext;

            middle->next = crr;
            middle = middleNext;

            crrNext = crrNext->next;
            middleNext = middleNext->next;
        }
    }
};

File: Linked-List/143-reorder-list/reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if(head == null || head.next == null || head.next.next == null){
            return;
        }

        ListNode slow = head;
        ListNode fast = head;
        while(fast.next != null && fast.next.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode crr = slow.next;
        ListNode prev = null;
        slow.next = null;

        while(crr != null){
            ListNode next = crr.next;
            crr.next = prev;
            prev = crr;
            crr = next;
        }

        ListNode first = head;
        ListNode second = prev;

        while(second != null){
            ListNode tmp1 = first.next;
            ListNode tmp2 = second.next;

            first.next = second;
            second.next = tmp1;

            first = tmp1;
            second = tmp2;
        }

    }
}